crates/oxc_diagnostics/src/service.rs

```diff
@@ -86,6 +98,53 @@ impl DiagnosticService {
         (path.to_path_buf(), diagnostics)
     }
 
+    /// # Panics
+    ///
+    /// * When the writer fails to write
+    pub fn json_output(&self) {
+        let mut buf_writer = BufWriter::new(std::io::stdout());
+
+        let mut messages_with_path = vec![];
+        while let Ok(Some((path, diagnostics))) = self.receiver.recv() {
+            let mut messages = vec![];
+            for diagnostic in diagnostics {
+                let severity = diagnostic.severity();
+                let is_error = severity.is_none() || severity == Some(Severity::Error);
+                let is_warning = severity == Some(Severity::Warning);
+                if is_warning || is_error {
+                    if is_warning {
+                        let warnings_count = self.warnings_count() + 1;
+                        self.warnings_count.set(warnings_count);
+                    }
+                    if is_error {
+                        let errors_count = self.errors_count() + 1;
+                        self.errors_count.set(errors_count);
+                    }
+                    // The --quiet flag follows ESLint's --quiet behavior as documented here: https://eslint.org/docs/latest/use/command-line-interface#--quiet
+                    // Note that it does not disable ALL diagnostics, only Warning diagnostics
+                    else if self.quiet {
+                        continue;
+                    }
+
+                    if let Some(max_warnings) = self.max_warnings {
+                        if self.warnings_count() > max_warnings {
+                            continue;
+                        }
+                    }
+                }
+
+                let message = JSONReportHandler::render_report(diagnostic.as_ref());
+                messages.push(message);
+            }
+            messages_with_path
+                .push(MessageWithPath { file_path: path.to_str().unwrap().to_owned(), messages });
+        }
+
+        let json_str = serde_json::to_string(&messages_with_path).unwrap();
+        self.output.set(json_str.clone());
+        buf_writer.write_all(json_str.as_bytes()).unwrap();
+    }
```

Comment:
I think we can DRY this by implementing a `ReportHandler` enum for `GraphicalReportHandler` and `JSONReportHandler`.

We can then configure the service with different reporters, what do you think?

Reply:
I agree, and I will process